### PR TITLE
feat: Template defaults

### LINKED ISSUE #3201: Allow setting defaults for `template` in workflows
[Edit — renamed to "Allow setting defaults for 'template' in workflows", and https://github.com/argoproj/argo-workflows/issues/6419 opened for `container`]

# Summary

Allow setting defaults for the `container` field in workflows

# Motivation

ref [Slack discussion](https://argoproj.slack.com/archives/C8J6SGN12/p1591723434019300)

Currently if we want to set a default image pull policy for containers, we need an additional layer of config, like kustomize, or to add it to every workflow manually.

# Proposal

I'm not sure; potentially allow this to flow through `workflowDefaults` like:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: workflow-controller-configmap
data:
  config: |
    workflowDefaults:
      metadata:
        annotations:
          argo: workflows
      spec:
        templates:
          container:
            imagePullPolicy: Always  # <- here
```

---

**Message from the maintainers**:

If you wish to see this enhancement implemented please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by sarabala1979**
Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).

closes #3201

### CODE REVIEW

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 1]
```diff
@@ -489,7 +492,7 @@ func (wfs *WorkflowSpec) HasPodSpecPatch() bool {
 // Template is a reusable and composable unit of execution in a workflow
 type Template struct {
 	// Name is the name of the template
-	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
```
Comment (sarabala1979) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:495]: This field is validated by CLI, UI, and controller.  But the only restriction when the user using swagger.json or SDK they need to set a dummy value on this.

**Comment by sarabala1979** [comment]
I will add more unit test to cover the code coverage

**Comment by alexec** [comment]
assigning @jessesuen for proto approval

**Comment by sarabala1979** [comment]
image
image

**Comment by sarabala1979** [comment]
image
image

File: docs/template-defaults.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+## Introduction
+
+`TemplateDefaults` feature enables the user to configure the default template values in workflow spec level that will apply to all the templates in the workflow. If the template has a value that also has a default value in `templateDefault`, the Template's value will take precedence. These values will be applied during the runtime.
+
+## Configuring `templateDefaults` in WorkflowSpec
+
+For example:
+```yaml
+apiVersion: argoproj.io/v1alpha1
```
Comment (alexec) [inline docs/template-defaults.md:12]: nice example

File: docs/template-defaults.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    retryStrategy: # retryStrategy value will be applied to all templates
+      limit: 2
+  templates:
+  - name: main
+    container:
+      image: docker/whalesay:latest
+```
+[template defaults example](examples/template_defaults.yaml)
+
+## Configuring `templateDefaults` in Controller level
```
Comment (alexec) [inline docs/template-defaults.md:29]: titlecase please

File: examples/template_defaults.yaml [hunk 4]
```diff
@@ -0,0 +1,40 @@
+apiVersion: argoproj.io/v1alpha1
```
Comment (alexec) [inline examples/template_defaults.yaml:1]: examples have hyphen is their filename

File: examples/template_defaults.yaml [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    - name: main
+      steps:
+        - - name: retry-backoff
+            template: retry-backoff
+        - - name: whalesay
+            template: whalesay
+
+    - name: whalesay
+      container:
+        image: docker/whalesay:latest
```
Comment (alexec) [inline examples/template_defaults.yaml:31]: should probably use argosay:v2

**Review by alexec — APPROVED** [review]
minor comments

**Pull request merged**